apps/studio/src/server/modules/audit/audit.service.ts

```diff
@@ -0,0 +1,181 @@
+import type {
+  AuditLogEvent,
+  Blob,
+  DB,
+  Resource,
+  ResourcePermission,
+  Transaction,
+  User,
+  VerificationToken,
+} from "../database"
+
+type FullResource = Resource & (Blob | undefined)
+
+interface CreateDelta {
+  before: null
+  after: FullResource
+}
+interface DeleteDelta {
+  before: FullResource
+  after: null
+}
+interface UpdateDelta {
+  before: FullResource
+  after: FullResource
+}
+
+interface ResourceEventLogProps {
+  eventType: Extract<
+    AuditLogEvent,
+    "ResourceCreate" | "ResourceUpdate" | "ResourceDelete" | "ResourceMove"
+  >
+  delta: CreateDelta | DeleteDelta | UpdateDelta
+  by: User
+  ip?: string
+}
+
+// NOTE: Type to force every logger method to have a tx
+type AuditLogger<T> = (tx: Transaction<DB>, props: T) => Promise<void>
+
+export const logResourceEvent: AuditLogger<ResourceEventLogProps> = async (
+  tx,
+  { eventType, delta, by, ip },
+) => {
+  await tx
+    .insertInto("AuditLog")
+    .values({ eventType, delta, userId: by.id, ipAddress: ip })
+    .execute()
+}
+
+interface LoginDelta {
+  before: VerificationToken
+  after: VerificationToken
+}
+
+// NOTE: logout just calls `session.destroy` and we only have
+// `userId` or sgid info in session.
+interface LogoutDelta {
+  before: User
+  after: null
+}
+
+interface AuthEventLogProps {
+  eventType: Extract<AuditLogEvent, "Login" | "Logout">
+  delta: LoginDelta | LogoutDelta
+  by: User
+  ip?: string
+}
+export const logAuthEvent: AuditLogger<AuthEventLogProps> = async (
+  tx,
+  { eventType, delta, by, ip },
+) => {
+  await tx
+    .insertInto("AuditLog")
+    .values({ eventType, delta, userId: by.id, ipAddress: ip })
+    .execute()
+}
+
+interface PublishEventLogProps {
+  by: User
+  delta: {
+    // NOTE: `null` if this is the first publish
+    // We don't want to store the `version` because it is a pointer
+    // to the blob/resource
+    // we will instead store the full data here so it is an accurate snapshot
+    before: (FullResource & { versionNumber: number }) | null
+    after: FullResource & { versionNumber: number }
+  }
+  eventType: Extract<AuditLogEvent, "Publish">
+  ip?: string
+}
+export const logPublishEvent: AuditLogger<PublishEventLogProps> = async (
```

Comment:
question: related to the site config, navbar and footer question above, how will we be recording publishes due to changes in those records?

Reply:
good question - i actually missed this out. lemme think abit more HAHA